### ISSUE TITLE
Enforce uniqueness of `display_order` on answers, within the scope of `question_id`

### ISSUE THREAD

**Issue by jefflunt** [NUARIG/surveyor#387]
It's possible to get inconsistently ordered answers in grids without enforcing uniqueness of the `display_order` on the `answer` model. As a workaround for now you can explicitly specify the `display_order` attribute in your survey DSL.
### Problem:

You have survey DSL that includes a grid, **and** the `display_order` attribute of answers is **not** specified in your DSL:

```
grid_severity "How strongly do you feel about the following, on a scale from 1-5, with 5 being 'most important'?" do 
  a "1"
  a "2"
  a "3"
  a "4"
  a "5"

  q "Quality"
  q "Presentation"
  q "Feel"
end
```

This problem can basically be avoided in the current surveyor if you specify a unique `display_order` in the DSL for **answers**, such as in this example:

```
grid_severity "How strongly do you feel about the following, on a scale from 1-5, with 5 being 'most important'?" do 
  a "1", :display_order => 1
  a "2", :display_order => 2
  a "3", :display_order => 3
  a "4", :display_order => 4
  a "5", :display_order => 5

  q "Quality"
  q "Presentation"
  q "Feel"
end
```

Without specifying a display order, and without enforcing uniqueness of this value, it's possible that multiple answers within a question grid will get the same `display_order` value. Since the Question `has_many :answers` association [specifies `display_order` as it default ordering](https://github.com/NUBIC/surveyor/blob/master/lib/surveyor/models/question_methods.rb#L10), and the `display_order` values are (potentially) non-unique, it's possible to get a grid where the **inputs** don't line up with the answer labels at the top, leading to users intending to select one answer, but the resulting `Response` record actually recording a different answer.

For example:

```
                1     2     3     4     5
Quality        (1)   (2)   (3)   (4)   (5)
Presentation   (2)   (3)   (4)   (5)   (1)   <= `answers` with the same `display_order` value returned in unexpected order
Feel           (1)   (2)   (3)   (4)   (5)
```

Depending on the DB your survey is sitting on, the order in which the `answer` records come out in this situation, when you call `questions.answers`, may or may not, in practice, be a consistent order. If they are **consistently wrong**, it's potentially fixable in your response data. If it's **inconsistently wrong**, then you're pretty much screwed, as you can't be sure that data collected with that grid is not, essentially, corrupted.
### Solution:

in `lib/surveyor/models/question_methods.rb`:

```
base.send :validates_uniqueness_of, :display_order, :scope => :question_id
```

...and a test to cover that.

**Comment by rsutphin** [comment]
Does the parser ensure that `display_order` is unique per question in current Surveyor versions? If not, then that will need to be changed before the proposed validation can be applied.

**Comment by jefflunt** [comment]
I'm not sure that it does, though importing a survey locally within my patched app (the above changes manually applied) seems to work. I'll definitely look at the parser as well in this context.

**Comment by jefflunt** [comment]
Would it be valuable to include in a patch a rake task that can check for existing, inconsistently ordered answers in grids, and/or grids that contain answers with non-unique `display_order` fields?

**Comment by yoon** [comment]
I definitely see a problem with having grid answers display out of order. But there are at least two ways to address this problem - make it easy (or default) to do the right thing, or make it impossible to do the wrong thing. I would be cautious in using rails validations, since they would make it impossible to save any answers that have non-unique display order attributes. Maybe that's what we want, but we should be sure that won't cause problems for existing users as well.

**Comment by jefflunt** [comment]
Sure, that's a reasonable question.

What's the use case for display orders not being unique? I realize that a uniqueness constraint on **all** answers is perhaps a bit of a ham-handed fix if the problem is only limited to grids, and I have noticed that the additional constraint does seem to significantly slow down the parsing process, since it has to check for scope uniqueness on every create. However, I guess my (perhaps naive) expectation is that answers will be rendered in surveyor in the order in which they appear in the DSL by default, unless I specify otherwise. Since render order is controlled by `display_order` already (at least according to the `default_scope`), it seems that having non-unique `display_order` attributes in a survey, for any question in any survey today, is likely to be accidental in all cases.

Considering other potential orders for answers: if, for example, someone wanted their answers to always be displayed in alphabetical/numeric/`id`/random order for whatever their application is, wouldn't they have to explicitly ask for that order today, similar to Levi's question on the dev list last week regarding the randomization of questions/answers?

I definitely get the inherent danger (and possible future technical debt) incurred by making such a broad-ranging change, but I'm also looking for where the harm is in this particular case. Granted, lack of proof of harm doesn't mean there isn't harm...

Is there a smaller footprint change that you think might be better?

**Comment by jefflunt** [comment]
I could see this maybe being an issue if someone had a survey editing tool that allowed existing surveys to be edited. I wouldn't want to break existing surveys or their ability to save changes. So, instead the validation of uniqueness could be on create only, while either doing nothing on an update, or issuing a warning.

**Comment by rsutphin** [comment]
> I have noticed that the additional constraint does seem to significantly slow down the parsing process, since it has to check for scope uniqueness on every create.

A change which would avoid the parser validation timing penalty would be a database-level unique index on `(question_id, display_order)` for answer. Since the parser would need to be guaranteeing that `display_order` was unique for the validation not to kill the parse anyway, and since survey structures are never changed (AFAIK) after parse, using an AR validation does not seem to have any benefits over a unique index.

However, that doesn't address @yoon's concern about people wanting to have manually-set `display_order`s where some of the values are the same. I can't think of any use cases for that — maybe we should ask on the mailing list to see if anyone has one?

> Is there a smaller footprint change that you think might be better?

The smaller-footprint change would be to change the way that grids are rendered so as to guarantee that the answer inputs were always ordered the same way for every question. It does seem like that would complicate grid rendering quite a bit, though. And I agree with this part:

> I guess my (perhaps naive) expectation is that answers will be rendered in surveyor in the order in which they appear in the DSL by default, unless I specify otherwise.

This seems to not only be your assumption, but the assumption of the code which renders the grid views also. It seems to me that guaranteeing this assumption at the model level is the safest way to fix the immediate grid rendering issue and also to forestall any future problems arising out of the same assumption.

**Comment by rsutphin** [comment]
> I could see this maybe being an issue if someone had a survey editing tool that allowed existing surveys to be edited.

@yipdw and I were talking about a similar concern related to the data model change he proposed with #386. In that case, I realized that a GUI survey editor that is based on updating existing Surveyor survey structure elements is a bad idea since individual survey elements are referenced from individual response records. Updating an answer's text, say, would destroy the semantics of any existing responses for that answer (and probably that whole question).

So, a GUI editor is going to have to take versioning into account and not just naively update surveys/sections/questions/answers.

**Comment by jefflunt** [comment]
> It seems to me that guaranteeing this assumption at the model level is the safest way to fix the immediate grid rendering issue and also to forestall any future problems arising out of the same assumption.

I agree with this, but isn't the scope to return answers by `display_order` already a model-level attempt to enforce a consistent rendering order? This seems to point back to the original issue, which is that this ordering is only **actually** consistent if the `display_order` values are unique. The fact that `display_order` values can be non-unique without direct intention **is the bug**. I'm not sure what other model-level solution would be both appropriate and non-redundant to the already implemented scope on the `answer` model.

The parser timing penalty is, I will admit, pretty severe. On a complex survey that I've been working with a lot lately, the parsing time is typically around 5-10 seconds. With the addition of the AR validation this kicks up to something like 60 seconds (6-10x more time consuming - yikes!). I will say that in my case surveys are imported/parsed on a seldom basis, and the total volume of surveys per week is relatively low. Because of this the speed impact worries me less than it would probably worry someone that does a lot of surveys all the time.

That being said, I certainly wouldn't mind the unique constraint to be placed on the database directly to avoid the penalty, if this also works for people on the mailing list.

> The smaller-footprint change would be to change the way that grids are rendered so as to guarantee that the answer inputs were always ordered the same way for every question. It does seem like that would complicate grid rendering quite a bit, though.

Yeah, I'm not so much a fan of complicating the rendering for a couple reasons.

First, that seems like a patch to fix an underlying data problem, and I'd rather the data was corrected than a view-level patch added to hide the data problem, because in order to enforce consistent ordering in the view you need to compare the text labels in the grid to the answer text, which duplicates the purpose of the `display_order` field. Also, any code that later deals with or concerns itself with rendering order would have to consider two factors (display order, and the corrective view code) instead of one (display order).

Second, a rendering-level fix has great potential danger for the data as well. If someone were to upgrade surveyor and didn't know that they had a broken grid before the upgrade, the rendering-level fix would actually complicate the task of fixing the response data. The change to rendering order in the view would create two versions of the response data: (1) new response data that contains the user-intended answer, and (2) exiting response data that contains the unintended, out of order answer.

Creating two bodies of response data whose only demarcation is the time at which surveyor was upgraded (to say nothing of multi-app server instances that share a database, but may not have the same app version/gem version on every app server at all times during an upgrade) could make fixing the existing response data incredibly difficult.

**Comment by kumbafu** [comment]
The parser timing penalty is, I will admit, pretty severe. On a complex survey that I've been working with a lot lately, the parsing time is typically around 5-10 seconds. With the addition of the AR validation this kicks up to something like 60 seconds (6-10x more time consuming - yikes!

Is there an index on that table?  If not, would anyone be opposed to creating one?

**Comment by rsutphin** [comment]
> Is there an index on that table? If not, would anyone be opposed to creating one?

There is not one. (There do not seem to be any indexes on `display_order`.) As I noted above, I'm in favor of using an index instead of the validation. But if the validation is going to be included, having an appropriate index also couldn't hurt.

**Comment by yoon** [comment]
> The fact that display_order values can be non-unique without direct intention is the bug.

Agreed. I'm just making the point that we need to fix it thoughtfully and with regard to the consequences.

Now that I've had a chance to look at the parser code, I realized that display_order is automatically set on answers in the parser. See: [parser.rb line 321](https://github.com/NUBIC/surveyor/blob/master/lib/surveyor/parser.rb#L321) and [parser.rb line 324](https://github.com/NUBIC/surveyor/blob/master/lib/surveyor/parser.rb#L324). What's actually happening is that the display order of all of the answers is getting set to 1.

**Comment by yoon** [comment]
> What's actually happening is that the display order of all of the answers is getting set to 1.

Sorry, that _was_ happening. It was fixed in 9ccb6030a4999a924596056eabfaed798fb76ce3 or before.

**Comment by yoon** [comment]
See #389
